ghcide/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -190,15 +194,22 @@ runLanguageServer options inH outH getHieDbLoc defaultConfig onConfigurationChan
             pure $ Right (env,ide)
 
 
+-- | Runs the action until it ends or until the given MVar is put.
+--   Rethrows any exceptions.
+untilMVar :: MonadUnliftIO m => MVar () -> m () -> m ()
```

Comment:
Kind of surprised these sorts of utilities aren't available elsewhere. I couldn't find them, anyway.

Reply:
Especially the pattern of forking a thread with an MVar to control terminating it :thinking:

Reply:
They may exist, I'm not sure either.